homeassistant/components/modbus/cover.py

```diff
@@ -0,0 +1,250 @@
+"""Support for Modbus covers."""
+import logging
+from typing import Any, Dict, Optional
+
+from pymodbus.exceptions import ConnectionException, ModbusException
+from pymodbus.pdu import ExceptionResponse
+import voluptuous as vol
+
+from homeassistant.components.cover import (
+    DEVICE_CLASSES_SCHEMA,
+    PLATFORM_SCHEMA,
+    SUPPORT_CLOSE,
+    SUPPORT_OPEN,
+    CoverEntity,
+)
+from homeassistant.const import CONF_DEVICE_CLASS, CONF_NAME, CONF_SLAVE
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from . import ModbusHub
+from .const import (
+    CALL_TYPE_COIL,
+    CALL_TYPE_REGISTER_HOLDING,
+    CALL_TYPE_REGISTER_INPUT,
+    CONF_COVERS,
+    CONF_HUB,
+    CONF_INPUT_TYPE,
+    CONF_REGISTER,
+    CONF_STATE_CLOSED,
+    CONF_STATE_CLOSING,
+    CONF_STATE_OPEN,
+    CONF_STATE_OPENING,
+    CONF_STATUS_REGISTER,
+    CONF_STATUS_REGISTER_TYPE,
+    DEFAULT_HUB,
+    DEFAULT_SLAVE,
+    MODBUS_DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```

Comment:
Sorry, I didn't think about this: We want the config to be hosted in the main integration for new platforms. What's the plan for that?

See https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md

Reply:
I didn't know about that, but I can rework the code. Could you provide some examples of platforms, which implement this new approach?

So, if I understand, the only difference is to move the config from `cover` section directly under modbus section, and to initialize covers from there. Something like:

```yaml
modbus:
  - name: plc_01
    type: tcp
    host: 127.0.0.1
    port: 5020
    covers:
        - name: plc_01_door
          hub: plc_01
          slave: 1
          device_class: gate
          coil: 0
          state_opening: 1
          state_open: 2
          state_closing: 3
          state_closed: 4
```

Reply:
Yeah, and then we need to load the cover platform via discovery with `load_platform`:

https://github.com/home-assistant/core/blob/f6df85f8cef4ac1b53eb66235f4f6a0c5acec513/homeassistant/helpers/discovery.py#L130-L137

The config can be passed either via `discovery_info` parameter or via `hass.data` to the platform.

Reply:
Thanks, I found a integration with this approach. Now I'm adapting the code. Documentation PR needs to be updated as well. I'll post an update here.